File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/ViewModel/FullScreenViewModel.java
package com.shuneault.netrunnerdeckbuilder.ViewModel;

import androidx.lifecycle.ViewModel;

import com.shuneault.netrunnerdeckbuilder.db.CardRepository;
import com.shuneault.netrunnerdeckbuilder.db.IDeckRepository;
import com.shuneault.netrunnerdeckbuilder.game.Card;
import com.shuneault.netrunnerdeckbuilder.game.Deck;
import com.shuneault.netrunnerdeckbuilder.helper.Sorter;

import java.util.ArrayList;
import java.util.Collections;

public class FullScreenViewModel extends ViewModel {

    private String setName;
    private String cardCode;
    private int position;
    private ArrayList<String> cardCodes;
    private IDeckRepository deckRepository;
    private Deck mDeck;
    private CardRepository cardRepository;
    private ArrayList<Card> mCards = new ArrayList<>();

    public FullScreenViewModel(IDeckRepository deckRepository, CardRepository cardRepository) {
        this.deckRepository = deckRepository;
        this.cardRepository = cardRepository;
    }

    public void setSetName(String setName) {
        this.setName = setName;
    }

    public String getSetName() {
        return setName;
    }

    public void setCardCode(String cardCode) {
        this.cardCode = cardCode;
        mCards.add(cardRepository.getCard(cardCode));
    }

    public String getCardCode() {
        return cardCode;
    }

    public void setPosition(int position) {
        this.position = position;
    }

    public int getPosition() {
        return position;
    }

    public void setCardCodes(ArrayList<String> cardCodes) {
        this.cardCodes = cardCodes;
        mCards.clear();
        mCards.addAll(cardRepository.getCards(cardCodes));
        // NO Card Sort
    }

    public ArrayList<String> getCardCodes() {
        return cardCodes;
    }

    public void loadDeck(long deckId) {
        mDeck = this.deckRepository.getDeck(deckId);
        mCards = mDeck.getCards();
        Collections.sort(mCards, new Sorter.CardSorterByCardType());
    }

    public Deck getDeck() {
        return mDeck;
    }

    public ArrayList<Card> getCards() {
        return mCards;
    }

    public String getFactionCode() {
        if (mDeck != null) {
            return mDeck.getIdentity().getFactionCode();
        }
        if (cardCode != null) {
            return mCards.get(0).getFactionCode();
        }
        return null;
    }
}

File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/ViewModel/FullScreenViewModel.kt
package com.shuneault.netrunnerdeckbuilder.ViewModel

import androidx.lifecycle.ViewModel
import com.shuneault.netrunnerdeckbuilder.db.CardRepository
import com.shuneault.netrunnerdeckbuilder.db.IDeckRepository
import com.shuneault.netrunnerdeckbuilder.game.Card
import com.shuneault.netrunnerdeckbuilder.game.CardCount
import com.shuneault.netrunnerdeckbuilder.game.Deck
import com.shuneault.netrunnerdeckbuilder.helper.Sorter
import com.shuneault.netrunnerdeckbuilder.helper.Sorter.CardSorterByCardType
import java.util.*
import kotlin.collections.ArrayList

class FullScreenViewModel(private val deckRepository: IDeckRepository, private val cardRepository: CardRepository) : ViewModel() {
    val cardCounts: ArrayList<CardCount> = ArrayList()
    var size: Int = 0
        get() = this.cardCounts.size

    var setName: String? = null
    private var cardCode: String? = null
    var position = 0
    // NO Card Sort
    var cardCodes: ArrayList<String> = ArrayList()
        set(cardCodes) {
            field.addAll(cardCodes)
            cardCounts.clear()

            for (code: String in cardCodes){
                cardCounts.add(CardCount(cardRepository.getCard(code), 0))
            }
            // NO Card Sort
        }
    private var deck: Deck? = null

    fun setCardCode(cardCode: String?) {
        cardCode?.let {
            this.cardCode = it
            cardCounts.add(CardCount(cardRepository.getCard(it), 0))
        }
    }

    fun loadDeck(deckId: Long) {
        deck = deckRepository.getDeck(deckId)
        updateCounts()
    }

    fun setDeck(deck: Deck){
        this.deck = deck
        updateCounts()
    }

    private fun updateCounts() {
        deck?.let {
            cardCounts.clear()
            cardCounts.addAll(it.cardCounts)
            cardCounts.sortWith(Sorter.CardCountSorterByTypeThenName());
            cardCounts.add(0, CardCount(it.identity, 1))
        }
    }

    fun getCurrentCardTitle(): String {
        val cc: CardCount = cardCounts[position]
        deck?.let {
            val num = it.getCardCount(cc.card);
            return String.format("%d x %s", num, cc.card.title)
        }
        return cc.card.title
    }

    fun getCurrentCard(): Card {
        val cc: CardCount = cardCounts[position]
        return cc.card
    }

    fun isEmpty(): Boolean {
        return cardCounts.isEmpty()
    }

    fun getDeckTitle(): CharSequence? {
        return deck?.name
    }

    val factionCode: String?
        get() {
            if (deck != null) {
                return deck!!.factionCode
            }
            return if (cardCode != null) {
                cardCounts[0].card.factionCode
            } else null
        }

}